Support for other halo finders

robjmcgibbon opened the pull request:
New version of https://github.com/SWIFTSIM/SOAP/pull/35. This PR merges the halo_finders branch with master. The reason for the new PR is because the COLIBRE branch has been merged into master since the halo_finders branch was created, and merging master into halo_finders created quite a few conflicts that needed fixing. 

It's might be helpful to compare the commits in this branch with those in halo_finders https://github.com/SWIFTSIM/SOAP/compare/halo_finders...SWIFTSIM:SOAP:merge_halo_finders

The main change is that the master branch and the halo_finders branch both had parameter files, but they weren't consistent with each other. This branch combines the two types of parameter files.

Any questions or comments please let me know. Sorry for letting the PR get to be this size, but hopefully now we're settled on HBT it will be easier to keep SOAP up-to-date.

### Other changes
- The documentation has been updated, with references to VR removed. Generating the pdf requires a parameter file to be passed, and InputHalos properties (e.g. InputHalos/HBTplus/TrackId) are included in the resulting table.

- I have updated the tests, so all unit tests run successfully with the command  `./tests/run_all.sh`

- The scripts directory has been updated. I've removed the individual scripts and parameters files for each halo finder, but we can put them back in if needed.

- The new properties from https://github.com/SWIFTSIM/SOAP/pull/60 are included in this branch, along with the change requested in the review.

- A new property, MassFractionExternal, from https://github.com/SWIFTSIM/SOAP/pull/70 is included.

- A bug fix for the recently heated gas filter f13147c5e570f23a704c27cf7502c5706d8dce09

- Save more HBT properties to the SOAP catalogue

- Rename the halo-ids argument as halo-indices, which I believe is less confusing

- Remove read_radius and search_radius from output

- Update calculation of SubhaloRankByBoundMass to work with different halo finders. I am not currently saving HBTplus/Rank, but could easily do so if it was deemed useful. For HBT on the L1000N1800 HYDRO run I get disagreement between HBTplus/Rank and SOAP/SubhaloRankByBoundMass for .3% of halos. I think this is due to how duplicate particles are handled differently in SOAP vs HBT, as can been seen from the following plot of bound mass. For the same reason I also get disagreement between SOAP/MaximumCircularVelocitySoft and HBTplus/VmaxPhysical. Hopefully setting orphan particle lists to be zero within HBT will reduce these differences.

- Move xray table definition to parameter file, and use new tables

<img src="https://github.com/SWIFTSIM/SOAP/assets/36136863/1a57d7d5-ee91-4aa9-9a7b-4685d09a61fa" width="400" height="300">

### Compare with master
This branch will not give identical results to master, so I had to make the following changes to allow for a comparison
- The first reason is because we now use the number of bound particles for the filters rather than the number of 6D FOF particles. To run master with the same filters I made the following changes https://gist.github.com/robjmcgibbon/3be2f8378a05e23b9c4c369b39896f0b.
- Another change is the definition of halo centre. Previously we were using the particle with the minimum potential from VR, rather than the most bound, but I have changed this to be consistent with HBT and Subfind. Therefore the following changes need to be applied to this branch before comparing with master https://gist.github.com/robjmcgibbon/fe7eb8364b9ffebf65d72841453f8900.
- The resulting outputs will still not agree due to the different output file format, and because of some bugfixes and changes mentioned above.  I compared output files using the following python script, https://gist.github.com/robjmcgibbon/2b2733675d2c911a255e48a768062610, which ignores fields that we expect to have changed.

With these changes I was happy that this branch is consistent with master for the L1000N0900 DMO_FIDUCIAL and HYDRO_FIDUCIAL VR catalogues at snapshots 17 & 77. Runtimes were roughly equal for both branches.

robjmcgibbon commented:
I need to run the formatter before merging to master

robjmcgibbon commented:
Plotting all objects from L1000N1800/HYDRO_FIDUCIAL from HBT run with the zero_particle_orphans branch
![hbt_mass](https://github.com/SWIFTSIM/SOAP/assets/36136863/1eb0b781-3b27-4b03-8316-8982394ee0ac)

jchelly commented:
Sorry for the delay. I've tried to run this and it all looks good so far. I'm working my way through your comparison link above now.

I see you removed use of deprecated unyt functions including uconcatenate. I used uconcatenate because with the unyt version we had at the time np.concatenate stripped the units. We might need to enforce a minimum unyt version to avoid wrong results.

jchelly commented on compute_halo_properties.py:
```diff
@@ -29,7 +29,7 @@
 import halo_properties
```
A lot of these are probably my fault, but there are various imports here we don't need any more: sys, traceback, h5py, phdf5, psort, halo_properties.

robjmcgibbon replied:
I ran https://github.com/PyCQA/autoflake to remove unused imports

robjmcgibbon commented:
I've just added a requirements file, but the python_env script remains for installing parallel h5py and the VirgoDC library

jchelly commented on match_vr_halos.py:
```diff
@@ -260,6 +260,61 @@ def consistent_match(match_index_12, match_index_21):
     # If we retrieved our own halo index, we have a match
     return np.where(match_back == local_halo_index, 1, 0)
 
+def get_match_vr_halos_args(comm):
```
We'll need to remove the import of the removed command_line_args file and change the call to get_match_vr_halos_args() to reference this version instead of the deleted file.

jchelly reviewed (commented):
This all looks good to me, and nicely tidied up and documented compared to how I left it. I've left some comments but they're mostly minor.

robjmcgibbon commented on read_hbtplus.py:
```diff
@@ -0,0 +1,354 @@
+    del halo_offset
+    del halo_index
+
+    # Now check for duplicates. HBTplus can assign the same particle to multiple
```
I would be inclined to convert it to a check, just in case we encounter any duplicates in the 10K run from a mechanism we haven't thought of yet. 

Should we get rid of the halo sizes file now we don't have duplicates?

jchelly replied:
I think so. We should be able to just use Nbound from the HBT output now.

robjmcgibbon replied:
I've removed the halo sizes in https://github.com/SWIFTSIM/SOAP/pull/71/commits/a113035145cc655af6a8465a43895218023d6276

robjmcgibbon replied:
I've stripped out the code that assigned duplicate particles to a single subhalo, and added an assert to check we don't have duplicates in https://github.com/SWIFTSIM/SOAP/pull/71/commits/019da21267f978bdac47c8b3f106b22ba303b68c

robjmcgibbon commented on read_hbtplus.py:
```diff
@@ -0,0 +1,354 @@
+        subhalo["NestedParentTrackId"][keep], units=unyt.dimensionless, dtype=int, registry=registry
+    )
+    m_bound = unyt.unyt_array(
+        subhalo["Mbound"][keep], units=unyt.dimensionless, dtype=float, registry=registry
```
Nope good catch. I don't think we need to save this now (same for vmax and rank), since they should be the same as the SOAP values. I'll test a couple of runs to make sure this is definitely true though.

robjmcgibbon commented on match_vr_halos.py:
```diff
@@ -260,6 +260,61 @@ def consistent_match(match_index_12, match_index_21):
     # If we retrieved our own halo index, we have a match
     return np.where(match_back == local_halo_index, 1, 0)
 
+def get_match_vr_halos_args(comm):
```
https://github.com/SWIFTSIM/SOAP/pull/71/commits/1b7f55032f07d42e83f6eeef121a7fec47575dbb

robjmcgibbon commented on SO_properties.py:
```diff
@@ -532,14 +573,23 @@ def ReducedTotalInertiaTensor(self):
+        Mass fraction contributed by particles that are bound to subhalos, but
+        are outside this FOF group. Includes particles from hostless subhalos.
+        """
+        return self.mass[self.is_bound_to_external].sum() / self.SO_mass
```
The bound mass here should only be particles within the SO radius, so I don't think it should ever be larger than the SO mass

robjmcgibbon commented on soap_args.py:
```diff
@@ -0,0 +1,128 @@
+    return args
+
+
+def get_match_vr_halos_args(comm):
```
Done in https://github.com/SWIFTSIM/SOAP/pull/71/commits/7be9761b163254ad8094e45eb4b54b61d86e2916

robjmcgibbon commented on scripts/COLIBRE/group_membership.sh:
```diff
@@ -23,26 +23,12 @@
-    ./vr_group_membership.py ${snapshot_filename} ${vr_basename} ${outfile}
+mpirun python3 -u -m mpi4py ./group_membership.py \
+       parameter_files/COLIBRE.yml \
+       --sim-name=${sim} --snap-nr=${snapnum}
```
Removed in https://github.com/SWIFTSIM/SOAP/pull/71/commits/8389407426649a736a080e555352e53603bf9dfa

robjmcgibbon commented on scripts/COLIBRE/halo_properties.sh:
```diff
@@ -13,51 +13,26 @@
-    --extra-input=${membership_filename} \
-    --parameters parameter_files/colibre_SOAP_params.yml
+       parameter_files/COLIBRE.yml \
+       --sim-name=${sim} --snap-nr=${snapnum} --chunks=1 ${dmo_flag}
```
Removed in https://github.com/SWIFTSIM/SOAP/pull/71/commits/8389407426649a736a080e555352e53603bf9dfa

robjmcgibbon commented on property_table.py:
```diff
@@ -2798,20 +2839,270 @@ class PropertyTable:
+            3,
+            np.float64,
+            "cMpc",
+            "Centre of potential. Used as reference for all relative positions. Equal to the position of the most bound particle in the subhalo.",
```
I agree. I changed it as part of https://github.com/SWIFTSIM/SOAP/pull/74, as well as a few of the other InputHalo property names

jchelly replied:
Ok, in that case there's no need to do anything here.

jchelly commented:
A couple of dataset names in property_table.py have underscores escaped with backslashes. If I delete my `__pycache__` directory then running the tests with tests/run_tasts.sh fails with 
```
SyntaxError: invalid escape sequence '\_'
```
If I have previously imported property_table.py so that it's cached the test succeeds! The backslashes only seem to be a problem for pytest for some reason.

Are the backslashes for latex? If so, maybe the names should be raw strings.

robjmcgibbon commented:
> If I have previously imported property_table.py so that it's cached the test succeeds! The backslashes only seem to be a problem for pytest for some reason.
> 
> Are the backslashes for latex? If so, maybe the names should be raw strings.

Yep it's for the documentation generation. I've pushed https://github.com/SWIFTSIM/SOAP/pull/71/commits/9b1411241edcfed5a814766e34b52e0dbdb5b16f following your suggestion, which allows the tests to run. However, I also updated these names in the other PR when I changed `cofp`

jchelly commented on parameter_files/FLAMINGO.yml:
```diff
@@ -0,0 +1,481 @@
+# Values in this section are substituted into the other sections
+Parameters:
+  sim_dir:     /cosma8/data/dp004/flamingo/Runs/
+  output_dir:  /cosma8/data/dp004/FLAMINGO/ScienceRuns/
```
I think we should certainly set the input directory since that probably wont change.

I just tried to run this branch on L1000N0900/DMO_FIDUCIAL and already forgot that the output_dir and scratch_dir are just base directories that don't need to include the simulation name. Having an example path in the parameter file might have helped with that.

robjmcgibbon replied:
I've added back in some example output directories (under my username, so no-one else will be able to write to them) in https://github.com/SWIFTSIM/SOAP/pull/71/commits/546f95caef21e8fbf7fb3a9ea77069b6c394dd06. I've also put in a check to make sure the output can be written to. This stops someone running all SOAP and then failing to write the output file. I haven't put in a check for the membership files since that's pretty cheap to run anyway.

jchelly commented on read_hbtplus.py:
```diff
@@ -0,0 +1,331 @@
+    snapshot_birth = unyt.unyt_array(
+        subhalo["SnapshotIndexOfBirth"][keep], units=unyt.dimensionless, dtype=int, registry=registry
+    )
+    parent_id = unyt.unyt_array(
```
Saving DescendantTrackId would enable construction of merger trees from just SOAP output, so it probably is useful.

robjmcgibbon replied:
Saving in https://github.com/SWIFTSIM/SOAP/pull/71/commits/f73baeeedbc29f3557b1c51bfd62abf5bb19d5a5

jchelly commented:
I've just tried running this branch on L1000N0900/HYDRO_FIDUCIAL to generate some input for testing the halo lightcone code. I made a few small changes to the flamingo scripts which I've put in #77.

robjmcgibbon commented on SO_properties.py:
```diff
@@ -2234,10 +2243,14 @@ def calculate_concentration(self, r):
 
     @lazy_property
     def concentration(self):
+        if self.skip_concentration:
```
Following some discussion with Joop & Matthieu today, we have decided to skip calculating concentration for certain SO definitions.

robjmcgibbon commented:
I'm going on to be on holiday from tomorrow until next Wednesday, so putting a list here of things I want to get done today so that the trimmed version of SOAP can be run on FLAMINGO-10k. I will still be checking slack/emails occasionally when I'm away. 

- [ ] I am going to close https://github.com/SWIFTSIM/SOAP/pull/74. This is because Joop would like us to filter halo types based on particle numbers rather than bound mass. I think it's better to use the pre-existing category_filter for this, rather than what I implemented in that PR. These changes aren't needed for the FLAMINGO-10k run anyway since for the trimmed down snapshots we aren't calculating the large physical apertures. As part of that PR I had renamed some of the `InputHalos` properties, so I will commit those changes directly to this branch.

- [ ] I am also going to close https://github.com/SWIFTSIM/SOAP/pull/43. I think we can just store REnclose (the maximum particle radius of each object) rather than 6 numbers. I will also commit that to this branch, since it's a single property which is very easy to compute.

- [ ] I need to make https://github.com/SWIFTSIM/SOAP/pull/75 ready for review

- [ ] Following some discussion we would like to change the COM and SORadius of halos to be comoving. PhD students have been using SORadius to load particles, but they all assume it is physical. I will add those changes to https://github.com/SWIFTSIM/SOAP/pull/78, since I already added a cMpc unit type there.

- [ ] Create a new property called HostHaloIndex, which will be stored within the SOAP group in the output file. This property will store the SOAP index of the central halo for each satellite. It is similar to `VR/HostHaloId`, except it stores the SOAP index rather than and ID. As part of this I would like to rename `InputHalos\index` (to make it explicit that this refers to the index of the input subhalo catalog), and also `HBTplus/HostHaloId` (to make it explicit that this property refers to FOF halos). I'll commit these changes to this branch.

It would be good to merge the following before running trimmed SOAP on FLAMINGO-10k. Hopefully they'll all be ready for review by the end of today.
- https://github.com/SWIFTSIM/SOAP/pull/72
- https://github.com/SWIFTSIM/SOAP/pull/75
- https://github.com/SWIFTSIM/SOAP/pull/78
- https://github.com/SWIFTSIM/SOAP/pull/79

VictorForouhar reviewed (approved):
Well done.
